### PR TITLE
Mixin Stability/Compat Overhaul and Conform to Mixin Standards

### PR DESCRIPTION
**Pull request by Bawnorton**
- Using loom 1.4 for VineFlower support (nicer decompiler for reading mc code)
  - Now using gradle 8.3
- Replace `@Redirect` calls that call the original method with `@WrapOperation` from MixinExtras (bundled in FLoader 0.15.0+) (allows chaining)
- Replace `@ModifyArgs` calls that target a single arg with`@ModifyArg` at `index`
- Replace cancellable `@Inject`s at `TAIL/RETURN` with `@ModifyReturnValue` from MixinExtras (allows changing).
- Remove local capture, local capture depends on the entire `LVT` being consistent (same vars and same order), instead it's better to rely on capturing single locals via `@Local` from MixinExtras, or capturing their usage in a `ThreadLocal`
- Removed enclosing method arguments from handlers that don't use them (where possible) to improve mixin stability
  - Less likely to need a re-write or change when the target code changes 
- Mixins are all abstract, even if they don't explicitly need to be, they should.
- Handler names should describe what they do rather than what they are, the only time you see them is in stack traces, so it's helpful to know exactly what broke and what it was trying to do.
- Annotate and prefix merged methods and fields with your modid to ensure there are no collisions (good practice).
- Annotate mixins that target classes which may not exist at runtime with `@Psuedo`
- Added back in a few features that I believe were unintentionally omitted during the update to 1.20.5 
  - [ItemOverridesMixin](https://github.com/Bawnorton/noxesium/blob/feature/1.20.5/fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/model/ItemOverridesMixin.java#L43)
  - [GuiHookMixin](https://github.com/Bawnorton/noxesium/blob/feature/1.20.5/fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/render/GuiHookMixin.java#L116-L124)

### CODE REVIEW

File: fabric/build.gradle.kts [hunk 1]
```diff
@@ -34,11 +35,13 @@ dependencies {
     api(project(":api"))
 
     // Compatibility with other mods
+    modCompileOnly("maven.modrinth:sodium:${property("sodium")}")
```
Comment (Aeltumn) [inline fabric/build.gradle.kts:38]: Reason for having enableSodium and enableIris was to allow for snapshot development, Sodium doesn't have a public snapshot version out so I set things up to be able to compile the game without.

File: fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/model/ItemOverridesMixin.java [hunk 2]
```diff
@@ -22,15 +21,26 @@
...
+
+    @Redirect(
+            method = "getItemOverrides",
+            at = @At(
+                    value = "FIELD",
+                    opcode = Opcodes.GETSTATIC,
+                    target = "Lnet/minecraft/client/renderer/block/model/ItemOverrides;EMPTY:Lnet/minecraft/client/renderer/block/model/ItemOverrides;"
+            )
+    )
+    public ItemOverrides replaceEmptyItemOverrides() {
```
Comment (Aeltumn) [inline fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/model/ItemOverridesMixin.java:43]: I intentionally removed this as vanilla's overrides when empty run no code anyway, so creating my own EMPTY object is just unnecessary.

File: fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/render/GuiHookMixin.java [hunk 3]
```diff
@@ -110,4 +112,14 @@ public void renderTabList(GuiGraphics graphics, float partialTicks, CallbackInfo
     public void renderSavingIndicator(GuiGraphics graphics, float partialTicks, CallbackInfo ci) {
         // TODO implement optimizations
     }
+
+    @WrapWithCondition(method = "*", at = @At(value = "INVOKE", target = "Lnet/minecraft/util/profiling/ProfilerFiller;push(Ljava/lang/String;)V"), require = 0)
+    private boolean ignoreProfilerPush(ProfilerFiller profilerFiller, String name) {
```
Comment (Aeltumn) [inline fabric/src/main/java/com/noxcrew/noxesium/mixin/performance/render/GuiHookMixin.java:117]: I would keep the profiler functional, previously it was kinda unintentionally removed as I had just replaced the whole method, but now with the new approach I dropped it. There's also no reason to remove them for any performance gains, and if anything any of the vanilla ones should just stay.

**Review by Aeltumn — CHANGES_REQUESTED** [review]
Interesting stuff, I did see MixinExtras is now bundled but I had no idea what features it had. I've left a few comments but generally these are good changes, thanks!

Semi-related but something I had on my own list to improve was to reduce access widener usage, afaik it's best practice to avoid using them. It's something I'd like to do to further improve compatibility in the future.

File: fabric/build.gradle.kts (hunk 1 above)
Comment (Bawnorton) [reply fabric/build.gradle.kts:38]: To compile the sodium-targeting mixins you would either need to include the class at compile time or reference the class via a target string instead. I'll swap the class references out for target strings so you don't need a compile-time reference.

Comment (Bawnorton) [reply fabric/build.gradle.kts:38]: nvm I see you have the conditions in the mixin plugin so they'll never run. I understand now

**Comment by Bawnorton** [comment]
You can find detailed information on mixin extras features on its GitHub wiki https://github.com/LlamaLad7/MixinExtras/wiki
It's quite powerful and definitely useful for compatibility

**Comment by Bawnorton** [comment]
> Semi-related but something I had on my own list to improve was to reduce access widener usage, afaik it's best practice to avoid using them. It's something I'd like to do to further improve compatibility in the future.

Yes that is correct, I did see you were using access wideners for methods and field but I figured I'd make a different pr to change that.

Access wideners create quite a lot of bloat in your dev env as every time you re-generate sources with changes made to the access widener you end up regenerating every class and it is cached as a separate jar, thus, you can easily end up with 10-20+ mc jars in your gradle cache, but you're only ever using 1.
They are also less clear to people reading your code as it's not clear that some field or method has been aw'd unless you look inside the access widener, the casting to an accessor mixin makes it obvious.

File: fabric/build.gradle.kts (hunk 1 above)
Comment (Aeltumn) [reply fabric/build.gradle.kts]: Yeah, I set up something for gradle to exclude the Sodium files to make it function. No release should be built without Sodium there, this is just for testing purposes.

**Pull request merged**